test/patest_wasapi_eac3.c

```diff
@@ -0,0 +1,214 @@
+                            unsigned long framesPerBuffer,
+                            const PaStreamCallbackTimeInfo* timeInfo,
+                            PaStreamCallbackFlags statusFlags,
+                            void *userData )
+{
+    paTestData *data = (paTestData*)userData;
+    short *out = (short*)outputBuffer;
+    unsigned long i,j;
+
+    (void) timeInfo; /* Prevent unused variable warnings. */
+    (void) statusFlags;
+    (void) inputBuffer;
+
+    /* stream out contents of data->buffer looping at end */
+
+    for( i=0; i<framesPerBuffer; i++ )
+    {
+        for( j = 0; j < CHANNEL_COUNT; ++j ){
+            *out++ = data->buffer[ data->playbackIndex++ ];
+
+            if( data->playbackIndex >= data->bufferSampleCount )
+                data->playbackIndex = 0; /* loop at end of buffer */
+        }
+    }
+
+    return paContinue;
+}
+
+/*******************************************************************/
+int main(int argc, char* argv[])
+{
+    PaStreamParameters outputParameters;
+    PaWasapiStreamInfo wasapiStreamInfo;
+    PaStream *stream;
+    PaError err;
+    paTestData data;
+    int deviceIndex;
+    FILE *fp;
+    const char *fileName = "d:\\test_48k.eac3.spdif";
+    data.buffer = NULL;
+
+    if( argc >= 2 )
+        fileName = argv[1];
+
+    printf( "reading spdif eac3 raw stream file %s\n", fileName );
+
+    fp = fopen( fileName, "rb" );
+    if( !fp ){
+        fprintf( stderr, "error opening spdif eac3 file.\n" );
+        return -1;
+    }
+    /* get file size */
+    fseek( fp, 0, SEEK_END );
+    data.bufferSampleCount = ftell( fp ) / sizeof(short);
+    fseek( fp, 0, SEEK_SET );
+
+    /* allocate buffer, read the whole file into memory */
+    data.buffer = (short*)malloc( data.bufferSampleCount * sizeof(short) );
+    if( !data.buffer ){
+        fprintf( stderr, "error allocating buffer.\n" );
+        return -1;
+    }
+
+    fread( data.buffer, sizeof(short), data.bufferSampleCount, fp );
+    fclose( fp );
+
+    data.playbackIndex = 0;
+
+    err = Pa_Initialize();
+    if( err != paNoError ) goto error;
+
+    deviceIndex = Pa_GetHostApiInfo( Pa_HostApiTypeIdToHostApiIndex(paWASAPI) )->defaultOutputDevice;
+    if( argc >= 3 ){
+        sscanf( argv[1], "%d", &deviceIndex );
+    }
+
+    printf( "using device id %d (%s)\n", deviceIndex, Pa_GetDeviceInfo(deviceIndex)->name );
+
+
+    outputParameters.device = deviceIndex;
+    outputParameters.channelCount = CHANNEL_COUNT;
+    outputParameters.sampleFormat = paInt16;
+    outputParameters.suggestedLatency = Pa_GetDeviceInfo( outputParameters.device )->defaultLowOutputLatency;
+    outputParameters.hostApiSpecificStreamInfo = NULL;
+
+    wasapiStreamInfo.size = sizeof(PaWasapiStreamInfo);
+    wasapiStreamInfo.hostApiType = paWASAPI;
+    wasapiStreamInfo.version = 1;
+    wasapiStreamInfo.flags = paWinWasapiExclusive | paWinWasapiUseChannelMask | paWinWasapiMatchEac3;
+    wasapiStreamInfo.channelMask = PAWIN_SPEAKER_5POINT1;
```

Comment:
Something is fishy here. `CHANNEL_COUNT` is 2 but you're using a 5.1 channel mask?

Reply:
@RossBencina I think AC3/EAC3 5.1 data is transferred inside 2-channel PCM stream because SPDIF does not have enough throughput for uncompressed 5.1 PCM. So in this sense it looks correct.

Reply:
@RossBencina @dmitrykos In this section, I refer to the VLC code, and its channel_count and channel_mask are assigned in this way. After testing, although the channel_count is set to 2, the output is still Dolby Atmos.

Reply:
@dmitrykos @gabys999 my question is really about whether this `wasapiStreamInfo.channelMask` must match the channel configuration that is encoded in the AC3/EAC3-SPDIF data stream or if it is unnecessary to specify a channel mask here (because the channel information is encoded in the SPDIF data stream). If it is working as-is I guess it does not matter, I'm just checking whether this code is redundant or unnecessary.

Reply:
Further, if it is necessary, it would be good to add a comment like `/* IMPORTANT: you must specify the correct channel mask here. */` or if it is not necessary use the comment: `/* Set channel mask for consistency with the SPDIFF data format, but it seems to work either way. */`